src/main/clojure/jarl/builtins/jwt.clj

```diff
@@ -0,0 +1,308 @@
+                     (str/replace "-----END PUBLIC KEY-----" "")
+                     (str/replace (System/lineSeparator) "")
+                     (base64/decode-bytes)
+                     (X509EncodedKeySpec.))
+        public-key (.generatePublic (KeyFactory/getInstance (if (es-alg? alg) "EC" "RSA")) key-spec)]
+    (reify VerificationKeyResolver
+      (resolveKey [_ _ _]
+        (if (es-alg? alg)
+          (cast ECPublicKey public-key)
+          (cast RSAPublicKey public-key))))))
+
+(defn x509-key-resolver [^String s]
+  (if-not (or (str/ends-with? s "-----END CERTIFICATE-----")
+              (str/ends-with? s "-----END CERTIFICATE-----\n"))
+    (throw (errors/builtin-ex "extra data after a PEM certificate block"))
+    (try
+      (let [der (strip-pem-wrapper s)]
+        (-> (X509Util.)
+            (.fromBase64Der der)
+            (List/of)
+            (X509VerificationKeyResolver.)
+            (doto (.setTryAllOnNoThumbHeader true))))
+      (catch Throwable _
+        (throw (errors/builtin-ex "failed to parse a PEM certificate"))))))
+
+(defn key-resolver-name [^String certificate]
+  (if (str/starts-with?     certificate "{")
+    (if (str/includes?      certificate "keys") "jwks" "jwk")
+    (if (str/starts-with?   certificate "-----BEGIN PUBLIC KEY-----")
+      "pem"
+      (if (str/starts-with? certificate "-----BEGIN CERTIFICATE-----")
+        "x509"
+        (throw (errors/builtin-ex "failed to extract a Key from the PEM certificate"))))))
+
+(defn key-resolver ^VerificationKeyResolver [certificate alg]
+  (condp = (key-resolver-name certificate)
+    "jwks" (jwks-key-resolver certificate)
+    "jwk"  (jwk-key-resolver  certificate)
+    "pem"  (pem-key-resolver  certificate alg)
+    "x509" (x509-key-resolver certificate)
+    (throw (errors/builtin-ex "no resolver found for %s" certificate))))
+
+(defn verify-hmac [jwt ^String secret]
+  (let [jws (doto (JsonWebSignature.)
+                  (.setDoKeyValidation false)
+                  (.setCompactSerialization jwt)
+                  (.setKey (HmacKey. (.getBytes secret "UTF-8"))))]
+    (.verifySignature jws)))
+
+(defn- verify-no-jwe [header]
+  (if (contains? header "enc")
+    (throw (errors/builtin-ex "JWT is a JWE object, which is not supported"))
+    header))
+
+(defn- strip-quotes [s]
+  (cond-> s (and (str/starts-with? s "\"") (str/ends-with? s "\"")) (subs 1 (dec (count s)))))
+
+(defn decode [jwt]
+  (let [parts       (split-jwt jwt)
+        header      (-> (first  parts) (base64-decode-bytes "header") (String.) json/read-str verify-no-jwe)
+        payload-str (-> (second parts) (base64-decode-bytes "payload") (String.))
+        sig         (-> (last   parts) (base64-decode-bytes "signature") hex/encode-bytes)]
+    (if (= "JWT" (get header "cty"))
+      (decode (strip-quotes payload-str)) ; nested JWT
+      [header (json/read-str payload-str) sig])))
+
+(defn verify-sig [jwt certificate alg]
```

Comment:
`certificate` is something of a misnomer. Consider `key` or `encoded-key`.

Reply:
Indeed, but it follows the naming used for the OPA built-ins: https://www.openpolicyagent.org/docs/latest/policy-reference/#tokens

Could still be worth changing here, but even if not entirely right it might make it easier to follow 🤔